astropy/coordinates/sky_coordinate.py

```diff
@@ -484,6 +487,94 @@ def transform_to(self, frame, merge_attributes=True):
+                             'position.')
+
+        if 'obstime' in self.frame.frame_attributes:
+            raise NotImplementedError("Updating the coordinates in a frame "
+                                      "with explicit time dependence is "
+                                      "currently not supported. If you would "
+                                      "like this functionality, please open an "
+                                      "issue on github:\n"
+                                      "https://github.com/astropy/astropy")
+
+        if new_obstime is not None and self.obstime is None:
+            # If no obstime is already on this object, raise an error if a new
+            # obstime is passed: we need to know the time / epoch at which the
+            # the position / velocity were measured initially
+            raise ValueError('This object has no associated `obstime`. '
+                             'apply_space_motion() must receive a time '
+                             'difference, `dt`, and not a new obstime.')
+
+        t1 = self.obstime
+        if t1 is None:
+            # MAGIC NUMBER: if the current SkyCoord object has no obstime,
+            # assume J2000 to do the dt offset.
+            t1 = Time('J2000')
+
+        if dt is None:
+            dt = new_obstime - self.obstime
+        t2 = t1 + dt
+
+        icrs = self.icrs
+        icrs.set_representation_cls(s=SphericalDifferential)
+
+        try:
+            plx = icrs.distance.to_value(u.arcsecond, u.parallax())
+        except u.UnitConversionError: # No distance: set to 0 by starpm convention
+            plx = 0.
+
+        try:
+            rv = icrs.radial_velocity.to_value(u.km/u.s)
+        except u.UnitConversionError: # No RV
+            rv = 0.
+
+        # proper motion in RA should not include the cos(dec) term, see the
+        # erfa function eraStarpv, comment (4).
+        starpm = erfa.starpm(icrs.ra.radian, icrs.dec.radian,
+                             icrs.pm_ra.to_value(u.radian/u.yr),
+                             icrs.pm_dec.to_value(u.radian/u.yr),
+                             plx, rv,
+                             2400000.5, t1.tdb.mjd,
+                             2400000.5, t2.tdb.mjd)
+
+        icrs2 = ICRS(ra=u.Quantity(starpm[0], u.radian, copy=False),
+                     dec=u.Quantity(starpm[1], u.radian, copy=False),
+                     pm_ra=u.Quantity(starpm[2], u.radian/u.yr, copy=False),
+                     pm_dec=u.Quantity(starpm[3], u.radian/u.yr, copy=False),
+                     distance=Distance(parallax=starpm[4] * u.arcsec, copy=False),
+                     radial_velocity=u.Quantity(starpm[5], u.km/u.s, copy=False),
+                     differential_cls=SphericalDifferential)
+
+        # Update the obstime of the returned SkyCoord, and need to carry along
+        # the frame attributes
+        # TODO: HACK for now is passing in .data because the SkyCoord init dies
```

Comment:
This is no longer necessary due to #6944 now being fixed to accept SphericalDifferential-holding frames, right?